README.md
# TerminalFormatters
Terminal font style and base 16 colors formatters


CHANGELOG.rst
v0.2.1
~~~~~~~~~

- Initial stable release

v0.2.1rc2
~~~~~~~~~

- Fix of syntax errors or undefined names according to the flake8 report:

.. code-block::

  0
  ./src/test_termformatters.py:113:9: F841 local variable 'S' is assigned to but never used
  ./src/test_termformatters.py:114:9: F841 local variable 'FG' is assigned to but never used
  ./src/test_termformatters.py:115:9: F841 local variable 'BG' is assigned to but never used
  ./src/termformatters/__init__.py:30:31: W605 invalid escape sequence '\]'
  ./src/termformatters/__init__.py:109:1: F401 'sys' imported but unused
  ./src/termformatters/__init__.py:112:1: F401 '.base.SelectFormatter' imported but unused
  ./src/termformatters/__init__.py:112:1: E402 module level import not at top of file
  ./src/termformatters/__init__.py:113:1: F401 '.base.BaseFormat' imported but unused
  ./src/termformatters/__init__.py:113:1: E402 module level import not at top of file
  ./src/termformatters/__init__.py:116:1: E402 module level import not at top of file
  ./src/termformatters/__init__.py:117:1: E402 module level import not at top of file
  ./src/termformatters/__init__.py:118:1: E402 module level import not at top of file
  ./src/termformatters/__init__.py:163:5: F841 local variable 'stylesList' is assigned to but never used
  ./src/termformatters/__init__.py:214:1: E266 too many leading '#' for block comment
  ./src/termformatters/base.py:3:1: F401 'sys' imported but unused
  ./src/termformatters/base.py:6:1: F401 'functools' imported but unused
  ./src/termformatters/base.py:7:1: F401 'functools.wraps' imported but unused
  1     E266 too many leading '#' for block comment
  5     E402 module level import not at top of file
  6     F401 'sys' imported but unused
  4     F841 local variable 'stylesList' is assigned to but never used
  1     W605 invalid escape sequence '\]'
  17

- Changes in '__init__.py' doc string and comments according to final flake8 report:

.. code-block::

  0
  ./src/termformatters/__init__.py:30:31: W605 invalid escape sequence '\]'
  ./src/termformatters/__init__.py:211:1: E266 too many leading '#' for block comment
  1     E266 too many leading '#' for block comment
  1     W605 invalid escape sequence '\]'
  2

v0.2.1rc1
~~~~~~~~~

- Initial stable pre-release


src/test_termformatters.py
"""UnitTest for base formatters"""
import sys

# Ref: https://docs.python.org/3/library/unittest.html
import unittest

try:
    from termformatters import StyleFormatters
    from termformatters import ForegroundFormatters
    from termformatters import BackgroundFormatters
except ImportError as importError:
    print("Formatter import error!")
    print(importError)
    sys.exit(1)


class TestFormatters(unittest.TestCase):
    def setUp(self):
        """Definition of test variables"""

        self.prefixStart = "\x1b["
        self.prefixEnd = "m"
        self.suffix = "\x1b[0m"
        self.string = "Test"

        self.styleFormats = {
            1: "bold",
            2: "faint",
            3: "italic",
            4: "underline",
            5: "blink",
            7: "negative",
            8: "concealed",
        }

        self.colorFormatsFG = {
            30: "fg_black",
            31: "fg_red",
            32: "fg_green",
            33: "fg_yellow",
            34: "fg_blue",
            35: "fg_magenta",
            36: "fg_cyan",
            37: "fg_white",
            90: "fg_black_bright",
            91: "fg_red_bright",
            92: "fg_green_bright",
            93: "fg_yellow_bright",
            94: "fg_blue_bright",
            95: "fg_magenta_bright",
            96: "fg_cyan_bright",
            97: "fg_white_bright",
        }

        self.colorFormatsBG = {
            40: "bg_black",
            41: "bg_red",
            42: "bg_green",
            43: "bg_yellow",
            44: "bg_blue",
            45: "bg_magenta",
            46: "bg_cyan",
            47: "bg_white",
            100: "bg_black_bright",
            101: "bg_red_bright",
            102: "bg_green_bright",
            103: "bg_yellow_bright",
            104: "bg_blue_bright",
            105: "bg_magenta_bright",
            106: "bg_cyan_bright",
            107: "bg_white_bright",
        }

    def test_instances(self):
        """Creating formatters instances"""

        # Formatters instances
        self.S = StyleFormatters()
        self.FG = ForegroundFormatters()
        self.BG = BackgroundFormatters()

        self.assertIsInstance(
            self.S, StyleFormatters, msg=f"Error creating instance: {self.S}"
        )
        self.assertIsInstance(
            self.FG, ForegroundFormatters, msg=f"Error creating instance: {self.FG}"
        )
        self.assertIsInstance(
            self.BG, BackgroundFormatters, msg=f"Error creating instance: {self.BG}"
        )

        # Deleting the test objects
        del self.S
        del self.FG
        del self.BG

    def test_style_formatter(self):
        """Formatters methods testing

        Composition of a fully formatted string by adding a prefix
        and a suffix: prefix + string + suffix, where

        prefix = "\x1b["
            + style | foreground | background
            +"m"

        string = "Test string"

        suffix = "\x1b[0m"
        """

        # Formatters instances
        print("\n")
        S = StyleFormatters()
        if S:
            self.assertIsInstance(
                S, StyleFormatters, msg="Error creating instance: {}".format(S)
            )
            print("Style Formatter: {}".format(S))
        FG = ForegroundFormatters()
        if FG:
            self.assertIsInstance(
                FG, ForegroundFormatters, msg="Error creating instance: {}".format(FG)
            )
            print("Foreground Formatter: {}".format(FG))
        BG = BackgroundFormatters()
        if BG:
            self.assertIsInstance(
                BG, BackgroundFormatters, msg="Error creating instance: {}".format(BG)
            )
            print("Background Formatter: {}".format(BG))

        print()
        self.allBaseFormats = {
            **self.styleFormats,
            **self.colorFormatsFG,
            **self.colorFormatsBG,
        }
        self.assertIs(type(self.allBaseFormats), dict, msg=None)

        for currentFormat in self.allBaseFormats:
            format = str(currentFormat)
            # Building formatter method name
            func = self.allBaseFormats[currentFormat]
            if func.startswith("fg_") or func.startswith("bg_"):
                printName = func[3:].replace("_", " ")
            else:
                printName = func

            # Instance selection
            if 1 <= currentFormat <= 8:
                funcName = "S." + func + "('" + self.string + "')"
            elif currentFormat in list(self.colorFormatsFG.keys()):
                funcName = "FG." + func + "('" + self.string + "')"
            elif currentFormat in list(self.colorFormatsBG.keys()):
                funcName = "BG." + func + "('" + self.string + "')"
            else:
                funcName = None
                self.assertIsNotNone(
                    funcName, msg=f"Invalid formatter name received: {funcName}"
                )

            expectedResult = (
                self.prefixStart + format + self.prefixEnd + self.string + self.suffix
            )
            obtainedResult = eval(funcName)
            print(
                "Current formater is {:<17s} formatted string   `{}`".format(
                    repr(printName), obtainedResult
                )
            )

            self.assertIsNotNone(obtainedResult)
            self.assertIs(type(obtainedResult), str)

            self.assertEqual(obtainedResult, expectedResult)


if __name__ == "__main__":
    # Print current Python version
    print(sys.version)
    print()
    unittest.main(verbosity=2)


src/demo.py
#!/usr/bin/env python3
import sys

from termformatters import StyleFormatters
from termformatters import ForegroundFormatters
from termformatters import BackgroundFormatters


# Print current Python version
print(sys.version)
print()

"""Formatters"""
S = StyleFormatters()
FG = ForegroundFormatters()
BG = BackgroundFormatters()


# CP437 (249 F9)
testChars249 = "∙∙∙"

stylesPrint = {
    1: "bold",
    2: "faint",
    3: "italic",
    4: "underline",
    5: "blink",
    7: "negative",
    8: "concealed",
}

colorNumbersFG = {
    30: "fg_black",  # Black
    31: "fg_red",  # Red
    32: "fg_green",  # Green
    33: "fg_yellow",  # Yellow
    34: "fg_blue",  # Blue
    35: "fg_magenta",  # Magenta
    36: "fg_cyan",  # Cyan
    37: "fg_white",  # White
    90: "fg_black_bright",  # Black Bright
    91: "fg_red_bright",  # Red Bright
    92: "fg_green_bright",  # Green Bright
    93: "fg_yellow_bright",  # Yellow Bright
    94: "fg_blue_bright",  # Blue Bright
    95: "fg_magenta_bright",  # Magenta Bright
    96: "fg_cyan_bright",  # Cyan Bright
    97: "fg_white_bright",  # White Bright
}

colorNumbersBG = {
    40: "bg_black",
    41: "bg_red",
    42: "bg_green",
    43: "bg_yellow",
    44: "bg_blue",
    45: "bg_magenta",
    46: "bg_cyan",
    47: "bg_white",
    100: "bg_black_bright",
    101: "bg_red_bright",
    102: "bg_green_bright",
    103: "bg_yellow_bright",
    104: "bg_blue_bright",
    105: "bg_magenta_bright",
    106: "bg_cyan_bright",
    107: "bg_white_bright",
}


"""Foreground/Background Table"""
print()
print(
    S.bold(
        FG.white_bright(
            "--- 8/16 base colors: SGR attributes 30..37 (90..97) and 40..47 (100..107) ---"
        )
    )
)
print()
# Colors table header
print("     color name", end="")
print("  dfl", end="")
print("   040   041   042   043   044   045   046   047", end="")
print("   100   101   102   103   104   105   106   107")
print()

for colorNumberFG in list(colorNumbersFG.keys()):
    currentColorNameFG = colorNumbersFG[colorNumberFG]  # "fg_name"
    currentFGName = currentColorNameFG[3:].replace("_", " ")

    print("{:>15s}".format(currentFGName), end="")
    print(" ", end="")

    # Building FG method name
    # currentColorFuncFG = "FG." + currentColorNameFG + "(' " + "gYw" + " ')"
    currentColorFuncFG = "FG." + currentColorNameFG + "(' " + testChars249 + " ')"

    # Eval Foreground function with default background
    currentForegroundFunc = eval(currentColorFuncFG)

    print("{:^5s}".format(currentForegroundFunc), end="")
    print(" ", end="")

    for colorNumberBG in list(colorNumbersBG.keys()):
        currentColorNumberBG = str(colorNumberBG)

        # Getting the `Backgorund` method name
        currentColorNameBG = colorNumbersBG[colorNumberBG]  # "bg_name"

        # Building and Eval full function
        currentFullFuncName = (
            "BG." + currentColorNameBG + "(" + currentColorFuncFG + ")"
        )
        currentFullFunc = eval(currentFullFuncName)

        print("{:^5s}".format(currentFullFunc), end="")
        print(" ", end="")

    print()

print()

"""Styles Table"""
print()
print(S.bold(FG.white_bright("--- base styles: SGR attributes 1..7 ---")))
print()
# Styles table header
print("     style name", end="")
print("  stl", end="")
print("   000   001   002   003   004   005   006   007", end="")
print("   008   009   010   011   012   013   014   015")
print()

for style in list(stylesPrint.keys()):
    currentStyleName = stylesPrint[style]  # "style"

    # Building and Evla `Style` formatter
    currentStyleMethod = "S." + currentStyleName
    currentStyleFunc = "S." + currentStyleName + "('" + currentStyleName + "')"
    currentStyleEval = eval(currentStyleFunc)

    if currentStyleName != "concealed":
        # {:>[num + 8]s} format() padding value with eval()
        # look like 2 x ESC ("\1xb")
        print("{:>23s}".format(currentStyleEval), end="")
    else:
        # {:>[num]s} format() padding value
        print("{:>15s}".format(stylesPrint[style]), end="")

    print(" ", end="")
    # Printing Style Number
    # print(" {:^3s} ".format(str(style)), end="")
    formatedStyle = str("{:03d}".format(style))
    print(" {:3s} ".format(formatedStyle), end="")
    print(" ", end="")

    for colorNumberFG in list(colorNumbersFG.keys()):
        currentColorNameFG = colorNumbersFG[colorNumberFG]  # "fg_name"
        formatedColorNumber = str("{:03d}".format(colorNumberFG))

        # Building `Foreground` formatter method name
        currentColorFuncFG = (
            "FG."
            + currentColorNameFG
            + "('"
            + "{:3s}".format(formatedColorNumber)
            + "')"
        )

        # Full Style+Foreground formatter
        currentFullFuncName = currentStyleMethod + "(" + currentColorFuncFG + ")"
        currentFullFunc = eval(currentFullFuncName)

        print(" {:3s} ".format(currentFullFunc), end="")
        print(" ", end="")

    print()

print()


src/termformatters/base.py
"""Base 16 collors formatter
"""
# https://docs.python.org/3/library/functools.html
# import functools
from functools import update_wrapper

# https://docs.python.org/3/library/inspect.html
import inspect


class SelectFormatter:
    """Format decorator

    Attributes
    ----------

    func : object
        decorated function name
    suffix : str
        format reset string
    stylesSwitch : dict
        foreground style to style code definition
    colorsSwitchFG : dict
        foreground color to color code definition
    colorsSwitchBG : dict
        background color to color code definition

    Methods
    -------

    __init__
        decorator init
    __call__
        style/colors function wrapper
    """

    def __init__(self, func):
        """Decorator definition

        Parameters
        ----------

        func : object
            decorated function name

        """
        update_wrapper(self, func)
        self.func = func
        self.suffix = "\x1b[0m"
        # Styles
        self.stylesSwitch = {
            "bold": 1,
            "faint": 2,
            "italic": 3,
            "underline": 4,
            "blink": 5,
            "negative": 7,
            "concealed": 8,
        }
        # Foreground colors
        self.colorsSwitchFG = {
            "fg_black": 30,
            "fg_red": 31,
            "fg_green": 32,
            "fg_yellow": 33,
            "fg_blue": 34,
            "fg_magenta": 35,
            "fg_cyan": 36,
            "fg_white": 37,
            "fg_black_bright": 90,
            "fg_red_bright": 91,
            "fg_green_bright": 92,
            "fg_yellow_bright": 93,
            "fg_blue_bright": 94,
            "fg_magenta_bright": 95,
            "fg_cyan_bright": 96,
            "fg_white_bright": 97,
        }
        # Background colors
        self.colorsSwitchBG = {
            "bg_black": 40,
            "bg_red": 41,
            "bg_green": 42,
            "bg_yellow": 43,
            "bg_blue": 44,
            "bg_magenta": 45,
            "bg_cyan": 46,
            "bg_white": 47,
            "bg_black_bright": 100,
            "bg_red_bright": 101,
            "bg_green_bright": 102,
            "bg_yellow_bright": 103,
            "bg_blue_bright": 104,
            "bg_magenta_bright": 105,
            "bg_cyan_bright": 106,
            "bg_white_bright": 107,
        }

    def __call__(self, initString, *args, **kwargs):
        """Style/Colors function wrapper"""
        self.formats = {
            **self.stylesSwitch,
            **self.colorsSwitchFG,
            **self.colorsSwitchBG,
        }
        if self.func.__name__ in self.formats.keys():
            formatNumber = self.formats.get(self.func.__name__, None)
            if formatNumber:
                prefix = "\x1b[" + str(formatNumber) + "m"
            else:
                prefix = self.suffix

            # Update result string
            updatedString = prefix + initString + self.suffix
            return self.func(updatedString, *args, **kwargs)
        else:
            # When style/color not in lists return original string
            return self.func(initString, *args, **kwargs)


class BaseFormat(object):
    """Main class decorator"""

    def __init__(self, decorator):
        self.decorator = decorator

    def __call__(self, cls, *args, **kwargs):
        """Assigning a decorator to a method"""
        # for funcName, funcRef in inspect.getmembers(cls, inspect.isfunction):
        for methodName, methodRef in inspect.getmembers(cls, inspect.ismethod):
            setattr(cls, methodName, self.decorator(methodRef))

        return cls


"""Formatters Classes Methods definition options:

: basic

    def bold(string: str) -> str: return string

: classmethod

    @classmethod
    def italic(cls, string: str) -> str: return string

: staticmethod

    @staticmethod
    def green(string: str) -> str: return string

"""


@BaseFormat(SelectFormatter)
class StyleFormatters:
    """Set of text style formatters"""

    @classmethod
    def bold(cls, string: str) -> str:
        return string

    @classmethod
    def faint(cls, string: str) -> str:
        return string

    @classmethod
    def italic(cls, string: str) -> str:
        return string

    @classmethod
    def underline(cls, string: str) -> str:
        return string

    @classmethod
    def blink(cls, string: str) -> str:
        return string

    @classmethod
    def negative(cls, string: str) -> str:
        return string

    @classmethod
    def concealed(cls, string: str) -> str:
        return string


@BaseFormat(SelectFormatter)
class ForegroundFormatters:
    """Set of foregroud color formatters"""

    @classmethod
    def fg_black(cls, string: str) -> str:
        return string

    @classmethod
    def fg_red(cls, string: str) -> str:
        return string

    @classmethod
    def fg_green(cls, string: str) -> str:
        return string

    @classmethod
    def fg_yellow(cls, string: str) -> str:
        return string

    @classmethod
    def fg_blue(cls, string: str) -> str:
        return string

    @classmethod
    def fg_magenta(cls, string: str) -> str:
        return string

    @classmethod
    def fg_cyan(cls, string: str) -> str:
        return string

    @classmethod
    def fg_white(cls, string: str) -> str:
        return string

    @classmethod
    def fg_black_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_red_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_green_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_yellow_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_blue_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_magenta_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_cyan_bright(cls, string: str) -> str:
        return string

    @classmethod
    def fg_white_bright(cls, string: str) -> str:
        return string


@BaseFormat(SelectFormatter)
class BackgroundFormatters:
    """Set of backgroud color formatters"""

    @classmethod
    def bg_black(cls, string: str) -> str:
        return string

    @classmethod
    def bg_red(cls, string: str) -> str:
        return string

    @classmethod
    def bg_green(cls, string: str) -> str:
        return string

    @classmethod
    def bg_yellow(cls, string: str) -> str:
        return string

    @classmethod
    def bg_blue(cls, string: str) -> str:
        return string

    @classmethod
    def bg_magenta(cls, string: str) -> str:
        return string

    @classmethod
    def bg_cyan(cls, string: str) -> str:
        return string

    @classmethod
    def bg_white(cls, string: str) -> str:
        return string

    @classmethod
    def bg_black_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_red_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_green_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_yellow_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_blue_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_magenta_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_cyan_bright(cls, string: str) -> str:
        return string

    @classmethod
    def bg_white_bright(cls, string: str) -> str:
        return string


src/termformatters/__init__.py
"""Terminal font style and base 16 colors formatters

Simple Python package for producing colored and formated terminal
text.

Terminals traditionally  take an input  of bytes and display them
as white text on a black  background.  But if the  input contains
specific  CSI (Control  Sequence Introducer)  sequences  then the
terminal may alter certain  display properties of the  text, such
as  the  style or  color.  Using   styles  and colors  in console
applications  helps us analyze  information faster and   focus on
important parts of the displayed information.

The package creates ANSI escape  character sequences for changing
the   color  and  style   of the  font, based   on human-readable
parameters like black, green, italic etc.

ANSI Escape Codes for Terminal Graphics

The ANSI  escape code standard, formally adopted as ISO/IEC 6429,
defines a  series of  control   sequences.  Each  control sequenc
e begins  with a Control   Sequence Introducer (CSI), defined  as
an escape character followed immediately   by a bracket:  `ESC[`.
The ANSI  ASCII  standard  represents the escape ESC character by
the decimal number 27 (33 in octal, 1B in hexadecimal).  The `ESC
[` is followed by any number (including   none) of "parameter byt
es" in the range `0x30–0x3F`,  then  by any  number of "intermedi
ate bytes" in the  range `0x20–0x2F`,  then finally  by  a single
"final byte"  in the range `0x40–0x7E`.

All  common sequences just   use  the parameters as  a  series of
semicolon-separated numbers. Missing numbers are treated as 0 and
no parameters at all in `ESC[m` acts like a 0 reset code.

Select Graphic Rendition (SGR) parameters

The control  sequence `CSI n   m`, named Select Graphic Rendition
(SGR), sets display attributes. Several attributes can be  set in
the   same sequence,    separated  by  semicolons.   Each display
attribute remains  in effect until  a following occurrence of SGR
resets it. If no codes are given, `CSI m` is treated as `CSI 0 m`
(reset/normal).

The package uses only  basic parameters to control the font style
and  its  color. The most commonly used SGR parameters to control
the font style in the range  (0-8) and the 8 actual colors within
the ranges (30-37, 40-47, 90-97, 100-107) are defined by the ANSI
standard. The package put these  parameters together to  create a
full SGR command.

The original  specification only had 8 colors, and just gave them
names. The SGR parameters  30–37  selected the  foreground color,
while  40–47 selected the   background.  Quite  a   few terminals
implemented "bold" (SGR code 1) as a brighter color rather than a
different  font,  thus  providing 8 additional foreground colors.
Usually you could not get    these  as background  colors, though
sometimes  inverse video (SGR code 7) would allow that. Examples:
to get black letters on white background use `ESC[30;47m`, to get
red use `ESC[31m`, to   get  bright red use  ESC[1;31m.  To reset
colors to their defaults, use `ESC[39;49m` (not supported on some
terminals),  or reset all attributes with ESC[0m. Later terminals
added the  ability to directly   specify the "bright" colors with
90–97 and 100–107.

The package supports the parameters shown in the tables below:

Font Styles SGR parameters

     Name           : Style :
    .........................
    "Bold"          :   1   :
    "Faint"         :   2   :
    "Italic"        :   3   :
    "Underline"     :   4   :
    "Blink"         :   5   :
    "Negative"      :   7   :
    "Concealed"     :   8   :

Foreground/Background Colors SGR parameters

     Name           :  FG  :  BG  :
    ...............................
    "Black"         :  30  :  40  :
    "Red"           :  31  :  41  :
    "Green"         :  32  :  42  :
    "Yellow"        :  33  :  43  :
    "Blue"          :  34  :  44  :
    "Magenta"       :  35  :  45  :
    "Cyan"          :  36  :  46  :
    "White"         :  37  :  47  :
    "Black Bright"  :  90  :  100 :
    "Red Bright"    :  91  :  101 :
    "Green Bright"  :  92  :  102 :
    "Yellow Bright" :  93  :  103 :
    "Blue Bright"   :  94  :  104 :
    "Magenta Bright":  95  :  105 :
    "Cyan Bright"   :  96  :  106 :
    "White Bright"  :  97  :  107 :

Test   scripts   in    the     source  code    repository print
formatted/colored tables using supported   ANSI sequences.  The
package has no requirements other than the standard library.

References
    https://en.wikipedia.org/wiki/ANSI_escape_code
"""
__version__ = "0.2.1"
# Main Formatters import
from .base import StyleFormatters
from .base import ForegroundFormatters
from .base import BackgroundFormatters


__all__ = ["StyleFormatters", "ForegroundFormatters", "BackgroundFormatters"]


def update_formatters_classes(cls):
    """Adding dynamic methods to formatters classes"""

    # Main formatters
    formatsList = [
        "fg_black",
        "fg_red",
        "fg_green",
        "fg_yellow",
        "fg_blue",
        "fg_magenta",
        "fg_cyan",
        "fg_white",
        "fg_black_bright",
        "fg_red_bright",
        "fg_green_bright",
        "fg_yellow_bright",
        "fg_blue_bright",
        "fg_magenta_bright",
        "fg_cyan_bright",
        "fg_white_bright",
        "bg_black",
        "bg_red",
        "bg_green",
        "bg_yellow",
        "bg_blue",
        "bg_magenta",
        "bg_cyan",
        "bg_white",
        "bg_black_bright",
        "bg_red_bright",
        "bg_green_bright",
        "bg_yellow_bright",
        "bg_blue_bright",
        "bg_magenta_bright",
        "bg_cyan_bright",
        "bg_white_bright",
    ]

    # Main styles
    # (disabled, additional aliases for styles will not be created)
    #
    # stylesList = [
    #     "green",
    #     "bold",
    #     "faint",
    #     "italic",
    #     "underline",
    #     "blink",
    #     "negative",
    #     "concealed",
    # ]

    methodsList = [
        method
        for method in dir(cls)
        if method.startswith("fg_") is True or method.startswith("bg_") is True
    ]

    notInList = []
    # Building method aliases
    for formatName in formatsList:
        for methodName in methodsList:
            aliasesList = []
            if formatName == methodName:
                methodNameParts = methodName.strip().split("_")
                # Standart colors
                if len(methodNameParts) == 2:
                    aliasesList.append(methodNameParts[1])
                # Bright colors
                elif len(methodNameParts) == 3:
                    aliasesList.append(
                        methodNameParts[0]
                        + "_"
                        + methodNameParts[2]
                        + "_"
                        + methodNameParts[1]
                    )
                    aliasesList.append(methodNameParts[1] + "_" + methodNameParts[2])
                    aliasesList.append(methodNameParts[2] + "_" + methodNameParts[1])
                else:
                    print("Unknown method name format: {}".format(methodName))
            else:
                notInList.append(formatName)

            if len(aliasesList) > 0:
                for aliasName in aliasesList:
                    destName = str(cls.__name__) + "." + methodName
                    setattr(cls, aliasName, staticmethod(eval(destName)))

    return (cls, notInList)


# Adding method aliases to Foreground and Background classes
#
update_formatters_classes(ForegroundFormatters)
update_formatters_classes(BackgroundFormatters)
